### PR TITLE
Policy cookie banner

### PR DESCRIPTION
**Pull request by kiran-mastek**
Bespoke Privacy and Cookie Banner

Check for privacy cookie existence
Create Privacy and Cookie Banner in ISAPI
Position partial for correcty layout
Update HTML for correct CSS
Use app setting to set the consent cookie expiration

Story: [AB#7250](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/7250)
Task: [AB#14640](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/14640),[AB#14647](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/14647)

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<section data-test-id="cookie-banner" class="bc-c-cookies nhsuk-header--white nhsuk-u-padding-top-3" aria-labelledby="cookies" role="banner">
+    <div class="nhsuk-width-container">
+        <h2 class="nhsuk-heading-m">Cookies on the Buying Catalogue</h2>
+        <p class="hsuk-body">We've put small files called cookies on your device to make the Buying Catalogue website work.</p>
+        <p class="hsuk-body">You can read more about our <a href="https://digital.nhs.uk/about-nhs-digital/privacy-and-cookies">cookies and privacy policy</a>.</p>
```
Comment (RahulJha-mastek) [inline src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml:5]: Link should be redirecting to new privacy-and-cookies page `href="/privacy-policy"`

File: tests/NHSD.BuyingCatalogue.Identity.Api.UnitTests/Extensions/HttpRequestExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
+﻿using System.Threading.Tasks;
```
Comment (RahulJha-mastek) [inline tests/NHSD.BuyingCatalogue.Identity.Api.UnitTests/Extensions/HttpRequestExtensionsTests.cs:1]: I think we can remove namespace `System.Threading.Tasks` in not needed.

Comment (kiran-mastek) [reply tests/NHSD.BuyingCatalogue.Identity.Api.UnitTests/Extensions/HttpRequestExtensionsTests.cs:1]: removed

File: src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml (hunk 1 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml]: done

**Review by RahulJha-mastek — APPROVED** [review]

File: src/NHSD.BuyingCatalogue.Identity.Api/Controllers/ConsentController.cs [hunk 3]
```diff
@@ -46,5 +52,17 @@ public async Task<IActionResult> Index(ConsentViewModel model)
 
             return View("Error");
         }
+
+        [AllowAnonymous]
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Identity.Api/Controllers/ConsentController.cs:56]: There's currently no authorization applied to `ConsentController` so the `AllowAnonymous` attribute shouldn't be needed.

File: src/NHSD.BuyingCatalogue.Identity.Api/Controllers/ConsentController.cs [hunk 4]
```diff
@@ -46,5 +52,17 @@ public async Task<IActionResult> Index(ConsentViewModel model)
...
+        [HttpGet("/dismiss-cookie-banner")]
+        public IActionResult DismissCookieBanner()
+        {
+            Response.Cookies.Append(Cookies.BuyingCatalogueConsent, "true", new CookieOptions
+            {
+                Expires = DateTime.Now.Add(cookieExpiration.ConsentExpiration),
+            });
+
+            return Redirect(Request.GetTypedHeaders().Referer.ToString());
+        }
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Identity.Api/Controllers/ConsentController.cs:66]: The `ConsentController` was originally for handling consent in the context of identity/IdentityServer only. Ideally, I feel this method probably belongs in a separate `CookieConsent` controller. However, given that this application will soon be retired due to the architectural revamp I think we may as well leave this here.

File: src/NHSD.BuyingCatalogue.Identity.Api/Extensions/HttpRequestExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿using Microsoft.AspNetCore.Http;
+using NHSD.BuyingCatalogue.Identity.Common.Constants;
+
+namespace NHSD.BuyingCatalogue.Identity.Api.Extensions
+{
+    public static class HttpRequestExtensions
+    {
+        public static bool ShowCookieConsent(this HttpRequest httpRequest) =>
+            httpRequest != null && !httpRequest.Cookies.ContainsKey(Cookies.BuyingCatalogueConsent);
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Identity.Api/Extensions/HttpRequestExtensions.cs:9]: `!= null` should be `is not null` to be consistent with the style used elsewhere in the solution.

File: src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<section data-test-id="cookie-banner" class="bc-c-cookies nhsuk-header--white nhsuk-u-padding-top-3" aria-labelledby="cookies" role="banner">
+    <div class="nhsuk-width-container">
+        <h2 class="nhsuk-heading-m">Cookies on the Buying Catalogue</h2>
+        <p class="hsuk-body">We've put small files called cookies on your device to make the Buying Catalogue website work.</p>
+        <p class="hsuk-body">You can read more about our <a href="/privacy-policy">cookies and privacy policy</a>.</p>
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml:5]: I think you'll need to inject `PublicBrowseSettings` and use `@PublicBrowseSettings.BaseAddress` here – see the Consent view, for example.

**Review by bjss-chris — COMMENTED** [review]
Looks fine in principle.

Have you tested these changes in the cluster (given how strict IdentityServer is about URLs)?

File: src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml (hunk 6 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Identity.Api/Views/Shared/_CookieConsent.cshtml:5]: Good point - added model for base URL to be passed through to partial as settings injected in the calling view.

File: src/NHSD.BuyingCatalogue.Identity.Api/Extensions/HttpRequestExtensions.cs (hunk 5 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Identity.Api/Extensions/HttpRequestExtensions.cs:9]: corrected

File: src/NHSD.BuyingCatalogue.Identity.Api/Controllers/ConsentController.cs (hunk 3 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Identity.Api/Controllers/ConsentController.cs:56]: removed

**Review by bjss-chris — COMMENTED** [review]
Can you update the minor version number? Easiest way is probably just to create an empty commit with the message "+semver:minor".

Good idea using the PB dismiss-cookie URL.

I might have missed your response but I'm not sure if you confirmed whether or not you've tested these changes inside the cluster.

**Review by bjss-chris — APPROVED** [review]

**Pull request merged**